library/core/src/str/mod.rs

```diff
@@ -212,7 +208,81 @@ impl str {
             None => index == self.len(),
 
             // This is bit magic equivalent to: b < 128 || b >= 192
-            Some(&b) => (b as i8) >= -0x40,
+            Some(&b) => b.is_utf8_char_boundary(),
+        }
+    }
+
+    /// Finds the closest `x` not exceeding `index` where `is_char_boundary(x)` is `true`.
+    ///
+    /// This method can help you truncate a string so that it's still valid UTF-8, but doesn't
+    /// exceed a given number of bytes. Note that this is done purely at the character level
+    /// and can still visually split graphemes, even though the underlying characters aren't
+    /// split. For example, the emoji 🧑‍🔬 (scientist) could be split so that the string only
+    /// includes 🧑 (person) instead.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// #![feature(nearest_char_boundary)]
+    /// let s = "❤️🧡💛💚💙💜";
+    /// assert_eq!(s.len(), 26);
+    /// assert!(!s.is_char_boundary(13));
+    ///
+    /// let closest = s.nearest_char_boundary_max(13);
+    /// assert_eq!(closest, 10);
+    /// assert_eq!(&s[..closest], "❤️🧡");
+    /// ```
+    #[unstable(feature = "nearest_char_boundary", issue = "none")]
+    #[inline]
+    pub fn nearest_char_boundary_max(&self, index: usize) -> usize {
+        if index >= self.len() {
+            self.len()
+        } else {
+            let lower_bound = index.saturating_sub(3);
+            let new_index = self.as_bytes()[lower_bound..=index]
+                .iter()
+                .rposition(|b| b.is_utf8_char_boundary());
+
+            // SAFETY: we know that the character boundary will be within four bytes
+            unsafe { lower_bound + new_index.unwrap_unchecked() }
+        }
+    }
+
+    /// Finds the closest `x` not below `index` where `is_char_boundary(x)` is `true`.
+    ///
+    /// This method is the natural complement to [`nearest_char_boundary_max`]. See that method
+    /// for more details.
+    ///
+    /// [`nearest_char_boundary_max`]: str::nearest_char_boundary_max
+    ///
+    /// # Panics
+    ///
+    /// Panics if `index > self.len()`.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// #![feature(nearest_char_boundary)]
+    /// let s = "❤️🧡💛💚💙💜";
+    /// assert_eq!(s.len(), 26);
+    /// assert!(!s.is_char_boundary(13));
+    ///
+    /// let closest = s.nearest_char_boundary_min(13);
+    /// assert_eq!(closest, 14);
+    /// assert_eq!(&s[..closest], "❤️🧡💛");
+    /// ```
+    #[unstable(feature = "nearest_char_boundary", issue = "none")]
+    #[inline]
+    pub fn nearest_char_boundary_min(&self, index: usize) -> usize {
```

Comment:
I think there are slightly less awkward names, reading the min/max names is not very intuitive unfortunately. I'm thinking of something like "next char boundary" or similar. I'm also sorry for bringing bikeshedding into this.

Maybe something like this pair:

- `find_char_boundary` ?
- `rfind_char_boundary` ?

Possibly the methods should have another word in the end, like `at` or `from`? Or we use `next`/`prev` as "verbs", but the non-inclusive connotation of those is a problem, similar to what you mentioned.

Reply:
The bikeshedding is welcome; as I've said I haven't the slightest idea how to best name these.

`find` and `rfind` seem closer but I kind of want to emphasise that they're used for snapping to boundaries, rather than finding them.

Reply:
Maybe char_boundary_floor and char_boundary_ceil?

Reply:
Not bad :)